libbeat/processors/script/javascript/session.go

```diff
@@ -0,0 +1,243 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package javascript
+
+import (
+	"reflect"
+
+	"github.com/dop251/goja"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+)
+
+const (
+	logName = "processor.javascript"
+
+	registerFunction   = "register"
+	entryPointFunction = "process"
+	testFunction       = "test"
+)
+
+// Session is an instance of the processor.
+type Session interface {
+	// Runtime returns the Javascript runtime used for this session.
+	Runtime() *goja.Runtime
+
+	// Event returns a pointer to the current event being processed.
+	Event() Event
+}
+
+// Event is the event being processed by the processor.
+type Event interface {
+	// Cancel marks the event as cancelled such that it will be dropped.
+	Cancel()
+
+	// IsCancelled returns true if Cancel has been invoked.
+	IsCancelled() bool
+
+	// Wrapped returns the underlying beat.Event being wrapped. The wrapped
+	// event is replaced each time a new event is processed.
+	Wrapped() *beat.Event
+
+	// JSObject returns the Value that represents this object within the
+	// runtime.
+	JSObject() goja.Value
+
+	// reset replaces the inner beat.Event and resets the state.
+	reset(*beat.Event) error
+}
+
+// session is a javascript runtime environment used throughout the life of
+// the processor instance.
+type session struct {
+	vm             *goja.Runtime
```

Comment:
Looking at the doc

> Is it goroutine-safe?
> No. An instance of goja.Runtime can only be used by a single goroutine at a time. You can create as many instances of Runtime as you like but it's not possible to pass object values between runtimes.

IIRC, I think this might be fine because when we call publish on the client we actually have a mutex around the processor calls. But if we ever change the contract that could cause issue.

Reply:
If we change that assumption we'll have to take a look at each processor to make sure they are thread-safe.